### PR TITLE
Relax `Resource: Sync` bound

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

Allow non-`Sync` types to be used as resources, without having to resort to `NonSend<>`.

Currently, non-`Sync` types are conflated with non-`Send` types, which results in overly restrictive scheduling. Any system that uses non-`Send` types must be run on the main thread, which increases thread contention and prevents parallelization. This is unnecessary, as `Send + !Sync` resources could be accessed from any thread, as long as it's only one at a time.

## Solution

* Relax the bound `Resource: Sync`.
* `Resource` trait tracks whether each type is `Sync`. The scheduler ensures that shared access to non-`Sync` types does not occur.

## Example

```rust
// This type can't be shared between threads, but it *can* be sent.
#[derive(Resource, Deref)]
#[resource(is_sync = false)] // compile fails if you forget this
pub struct MyResource(std::cell::Cell<u32>);

fn my_system(x: Res<MyResource>) {
    x.set(x.get() + 1);
}
```

## Future Work

For the time being, systems with read-only access to a non-`Sync` resource are run on the main thread. This is overly restrictive: we will need to update our `Access` model to properly support non-`Sync` reads.

---

## Changelog

+ `Resource` now supports types that do not implement `std::marker::Sync`.

## Migration Guide

The `Resource` trait is now `unsafe` to implement manually. Use of the derive macro is recommended.

```rust
// Before
struct MyResource(...);
impl Resource for MyResource {}

// After
struct MyResource(...);
unsafe impl Resource for MyResource {
    // Undefined Behavior if this is wrong. 
    const IS_SYNC: bool = true;
}
```

Any generic code that relied on `Resource: Sync` will now need to specify the `Sync` bound. 

```rust
// Before:
fn my_system<T: Resource>(
    my_resource: Res<T>,
    q: Query<...>,
) {
    // Since we access `my_resource` across threads (because of par_for_each), we need it to be `Sync`.
    q.par_for_each(|...| {
        do_something_with(&my_resource);
    })
}

// After:
fn my_system<T: Resource + Sync>( // Just add this bound.
    ...
) {
    ...
}
```

### CODE REVIEW

**Comment by maniwani** [comment]
We can avoid having to specify the `Sync` bound if `Access` knows which `ComponentId` or `ArchetypeComponentId` are `Sync`.

**Comment by JoJoJet** [comment]
Should be ready for review now. Note that this PR is blocked by #6603 .

File: crates/bevy_tasks/src/task_pool.rs [hunk 1]
```diff
@@ -299,8 +299,11 @@ impl TaskPool {
                     break result;
                 };
 
-                self.executor.try_tick();
-                task_scope_executor.try_tick();
+                std::panic::catch_unwind(|| {
+                    executor.try_tick();
+                    task_scope_executor.try_tick();
+                })
+                .ok();
```
Comment (james7132) [inline crates/bevy_tasks/src/task_pool.rs:306]: These seem unrelated. Can you revert these changes?

File: crates/bevy_transform/src/systems.rs [hunk 2]
```diff
@@ -318,10 +318,7 @@ mod test {
         let mut temp = World::new();
         let mut app = App::new();
 
-        // FIXME: Parallel executors seem to have some odd interaction with the other
-        // tests in this crate. Using single_threaded until a root cause can be found.
-        app.add_stage("single", SystemStage::single_threaded())
-            .add_system_to_stage("single", transform_propagate_system);
+        app.add_system(transform_propagate_system);
```
Comment (james7132) [inline crates/bevy_transform/src/systems.rs:321]: These seem unrelated. Can you revert these changes?

File: crates/bevy_ecs/src/system/system_param.rs [hunk 3]
```diff
@@ -371,6 +378,13 @@ impl<'a, T: Resource> SystemParam for Res<'a, T> {
 // conflicts with any prior access, a panic will occur.
 unsafe impl<T: Resource> SystemParamState for ResState<T> {
     fn init(world: &mut World, system_meta: &mut SystemMeta) -> Self {
+        // If the resource can't be shared between threads, make sure it runs on the main thread.
+        // FIXME: We should be able to register access as 'exclusive reads' to prevent shared access
+        // without forcing the system to run on the main thread.
```
Comment (james7132) [inline crates/bevy_ecs/src/system/system_param.rs:383]: IMO, there's no point to merging this PR as is without this. You would otherwise just use a `NonSend` for the same effect.

You can treat a `Res<T: !Sync>` the same as `ResMut<T: !Sync>`. Move the `if !T::IS_SYNC` down below, and change the non-Sync branch to add a write instead.

File: crates/bevy_ecs/src/world/mod.rs [hunk 4]
```diff
@@ -1431,12 +1431,14 @@ impl World {
     /// The returned pointer must not be used to modify the resource, and must not be
     /// dereferenced after the immutable borrow of the [`World`] ends.
     ///
+    /// If the resource type is `!Send`, you must not share the returned pointer between threads.
```
Comment (james7132) [inline crates/bevy_ecs/src/world/mod.rs:1434]: This is incorrect. `!Send` means the value cannot be sent to another thread. Sharing the pointer, `&T`, with other threads is required by `Sync`.

File: crates/bevy_ecs/macros/src/component.rs [hunk 5]
```diff
@@ -8,20 +8,76 @@ pub fn derive_resource(input: TokenStream) -> TokenStream {
...
+    }
 
     let struct_name = &ast.ident;
     let (impl_generics, type_generics, where_clause) = &ast.generics.split_for_impl();
 
     TokenStream::from(quote! {
-        impl #impl_generics #bevy_ecs_path::system::Resource for #struct_name #type_generics #where_clause {
+        // SAFETY: The trait bounds on this impl guarantee that the non-`Sync` types will not have `IS_SYNC = true`.
+        unsafe impl #impl_generics #bevy_ecs_path::system::Resource for #struct_name #type_generics #where_clause {
+            const IS_SYNC: bool = #is_sync;
```
Comment (james7132) [inline crates/bevy_ecs/macros/src/component.rs:34]: These should definitely have some compile fail tests to ensure that this breaks when this contract is broken.

Separately, we should see if the compile failures are readable by normal users.

**Review by james7132 — CHANGES_REQUESTED** [review]

File: crates/bevy_ecs/src/system/system_param.rs (hunk 3 above)
Comment (JoJoJet) [reply crates/bevy_ecs/src/system/system_param.rs:383]: > IMO, there's no point to merging this PR as is without this. You would otherwise just use a NonSend for the same effect.

I want to keep this PR as small as possible, so it's easier to review. If this gets merged, I'll open a follow-up PR that makes the `Access` changes.

These changes are easily separable, so I see no reason to roll them up into one PR. Especially since the `Access` changes are just an optimization.

Also, this already allows `ResMut<T: !Sync>` to be used off of the main thread.

> You can treat a Res<T: !Sync> the same as ResMut<T: !Sync>. Move the if !T::IS_SYNC down below, and change the non-Sync branch to add a write instead.

This won't work, it will mess up ambiguity detection.

File: crates/bevy_tasks/src/task_pool.rs (hunk 1 above)
Comment (JoJoJet) [reply crates/bevy_tasks/src/task_pool.rs:306]: This is a temporary fix for #6603

File: crates/bevy_ecs/src/world/mod.rs (hunk 4 above)
Comment (JoJoJet) [reply crates/bevy_ecs/src/world/mod.rs:1434]: ah, right

File: crates/bevy_ecs/macros/src/component.rs (hunk 5 above)
Comment (JoJoJet) [reply crates/bevy_ecs/macros/src/component.rs:34]: I added a compile fail test. The error message isn't great, but it isn't any worse than before this PR.

I tried a few ways of providing a more descriptive error message, but they all caused problems in edge cases.

**Comment by JoJoJet** [comment]
This is no longer blocked by #6603.

File: crates/bevy_ecs/macros/src/component.rs (hunk 5 above)
Comment (JoJoJet) [reply crates/bevy_ecs/macros/src/component.rs:34]: After investigating further, the error message is actually good for concrete types. It shows up when you define the struct, and points you directly to the field that makes the struct `!Sync`.

It's less good for generic types, because the error doesn't occur until you try to use the resource.

```rust
// Concrete types:

#[derive(Resource)] // error message shows up here, right where we want it.
struct MyResource(Cell<u32>);
```

```rust
// Generic types:

#[derive(Resource)]
struct MyResource<T>(Cell<T>);

// The error message doesn't show up until here, which is un-ideal.
fn my_system(_: Res<MyResource<u32>>) {}
```

IMO this isn't perfect, but it's perfectly acceptable.

**Comment by JoJoJet** [comment]
I'm closing this in favor of #6864 -- this all seems a bit unnecessary when you could just wrap your `!Sync` types in a `SyncCell`.

**Comment by james7132** [comment]
For what it's worth, I do think `!Sync` components by themselves are valuable. `SyncCell` forces mutable access which may force users into triggering change detection, even when it's really just for read-only access. This is bypasssable, but it's not exactly great UX.

**Comment by hymm** [comment]
could we create our own wrapper that would only trigger change detection on mutable access to the inner value?
